contracts/vault-managers/CDPManager01.sol

```diff
@@ -0,0 +1,355 @@
+
+                if (usdpAmount != 0) {
+                    uint fee = vault.calculateFee(asset, msg.sender, usdpAmount);
+                    vault.chargeFee(vault.usdp(), msg.sender, fee);
+                    vault.repay(asset, msg.sender, usdpAmount);
+                }
+
+                vault.update(asset, msg.sender);
+
+                _ensurePositionCollateralization(asset, msg.sender);
+            }
+        }
+
+        // fire an event
+        emit Exit(asset, msg.sender, assetAmount, usdpAmount);
+
+        return usdpAmount;
+    }
+
+    /**
+      * @notice Repayment is the sum of the principal and interest
+      * @dev Withdraws collateral and repays specified amount of debt
+      * @param asset The address of the collateral
+      * @param assetAmount The amount of the collateral to withdraw
+      * @param repayment The target repayment amount
+      **/
+    function exit_targetRepayment(address asset, uint assetAmount, uint repayment) external returns (uint) {
+
+        uint usdpAmount = _calcPrincipal(asset, msg.sender, repayment);
+
+        return exit(asset, assetAmount, usdpAmount);
+    }
+
+    /**
+      * @notice Withdraws WETH and converts to ETH
+      * @param ethAmount ETH amount to withdraw
+      * @param usdpAmount The amount of USDP token to repay
+      **/
+    function exit_Eth(uint ethAmount, uint usdpAmount) public returns (uint) {
+        usdpAmount = exit(WETH, ethAmount, usdpAmount);
+        IWETH(WETH).transferFrom(msg.sender, address(this), ethAmount);
+        IWETH(WETH).withdraw(ethAmount);
+        (bool success, ) = msg.sender.call{value:ethAmount}("");
+        require(success, "Unit Protocol: ETH_TRANSFER_FAILED");
+        return usdpAmount;
+    }
+
+    /**
+      * @notice Repayment is the sum of the principal and interest
+      * @notice Withdraws WETH and converts to ETH
+      * @param ethAmount ETH amount to withdraw
+      * @param repayment The target repayment amount
+      **/
+    function exit_Eth_targetRepayment(uint ethAmount, uint repayment) external returns (uint) {
+        uint usdpAmount = _calcPrincipal(WETH, msg.sender, repayment);
+        return exit_Eth(ethAmount, usdpAmount);
+    }
+
+    // decreases debt
+    function _repay(address asset, address user, uint usdpAmount) internal {
+        uint fee = vault.calculateFee(asset, user, usdpAmount);
+        vault.chargeFee(vault.usdp(), user, fee);
+
+        // burn USDP from the user's balance
+        uint debtAfter = vault.repay(asset, user, usdpAmount);
+        if (debtAfter == 0) {
+            // clear unused storage
+            vault.destroy(asset, user);
+        }
+    }
+
+    function _ensurePositionCollateralization(address asset, address user) internal view {
+        // collateral value of the position in USD
+        uint usdValue_q112 = IOracleUsd(oracleRegistry.oracleByAsset(asset)).assetToUsd(asset, vault.collaterals(asset, user));
```

Comment:
We're ignoring `vault.oracleType[asset][user]`. Is it ok?

Reply:
Moreover, we're ignoring `vault.isOracleTypeEnabled` as well. However, this setting may block liquidations, so should be in sync with borrowing.

Reply:
Oracle registry should be in sync with `isOracleTypeEnabled`